README
Math-BigInt

Math::BigInt::Constant - arbitrarily big integer constants

INSTALLATION

To install this module, unpack the distribution file, and run the following
commands:

    perl Makefile.PL
    make
    make test
    make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for these modules with the
perldoc command.

    perldoc Math::BigInt::Constant

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Math-BigInt-Constant

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Math-BigInt-Constant

    CPAN Ratings
        http://cpanratings.perl.org/d/Math-BigInt-Constant

    Search CPAN
        http://search.cpan.org/dist/Math-BigInt-Constant

COPYRIGHT AND LICENCE

Copyright 2001-2007 Tels, L<http://bloodgate.com>.

Copyright 2016- Peter John Acklam L<pjacklam@online.no>.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


(C) ..--- ----- ----- .--- - ..--- by - . .-.. ... .-.-.-  All rights reserved.

Requires new version of Math::BigInt and Math::BigFloat as found on CPAN.

This package allows you to store constant Math::BigInts like this:

	use Math::BigInt::Constant;

	my $x = Math::BigInt::Constant->new(2);

	print "$x\n";			# okay, prints 2
	print $x+2,"\n";		# okay, prints 4
	print $x++,"\n";		# not okay, would modify $x

Please send me test-reports, your experiences with this and your ideas - I love
to hear about my work!

Tels <http://bloodgate.com/>


CHANGES
1.12 Not Released

 * Convert test files t/pod.t and t/pod_cov.t into t/release-pod.t and
   t/release-pod-coverage.t.

 * Better Makefile.PL.

 * The tests now require Test::More version 0.88 or newer.

 * Add test file t/release-test-version.t, t/release-portability.t, and
   t/release-cpan-changes.t.

 * New MANIFEST.SKIP file based on the template in the ExtUtils-Manifest
   distribution.

 * Format CHANGES according to CPAN::Changes::Spec.

1.11 2016-12-08 pjacklam

 * Add more tests, including tests for the from_hex(), from_oct(), from_bin(),
   and from_bytes() to confirm latest fix in Math-BigInt-1.999804 regarding

1.10 2016-12-03 pjacklam

 * Use ExtUtils::MakeMaker rather than Module::Install in Makefile.PL

 * Remove Build.PL. It is no longer supported.

 * Remove INSTALL file. The information is in the README file.

 * Better documentation.

 * Update README file.

 * Update BUGS file.

 * Better documentation in BUGS and SUPPORT sections in .pm file.

 * Remove author information in LICENSE file.

 * Inform that the TODO file is not up to date.

 * CHANGES file now has the same formatting as other Math::BigInt-related
   distributions.

1.09 2016-11-28 pjacklam

 * In Makefile.PL, add "#!perl" to the top, as it gives nice syntax
   highlighting in some editors.

 * In Makefile.PL, change "require 5.6.2" to "require 5.006001". Math::BigInt
   requires only Perl v5.6.1, not v5.6.2.

 * Use more of the functionality provided by Test::More.

 * Add a few more test.

 * "Modernise" t/pod.t and t/pod_cov.t.

 * Required version of Math::BigInt is 1.999802 (was 1.85).

 * Required version of Test::More is 0.47 (was 0.62).

 * Update bundled version of Module::Install to 1.17 (was 0.65).

1.08 2007-05-05 Tels 322 tests

 * require Math::BigInt v1.85

 * bundle and use Module::Install

 * add POD tests

 * add tests for: bnok(), bexp(), as_oct() etc.

1.07 2006-03-14 Tels 129 tests

 * fix failing tests due to 0 vs '' difference

1.06 2005-01-01 Tels 129 tests

 * require Math::BigInt v1.74

 * use like() in tests

 * write our own META.yml in Makefile.PL to fix License field

1.05 2004-01-12 Tels 103 tests

 * added bfround(), bround(), bfac(), bmodinv(), bmodpow() to tests

 * removed some double tests

 * added some more tests for allowed operations

 * small doc fixes and enhancements

 * converted test file to Test::More

1.04 2002-02-10 Tels 72 tests

 * broken out from Math::Big

 * added blog()/bfac()/band() to the testsuite

1.03 2001-06-29

 * fixed copy(); added tests for bgcd() and blcm()

1.02 2001-06-15

 * fixed documentation to match current version

 * added tests for band(), bior(), bxor(), bceil(), bfloor(), bnan() and binf()

1.01 2001-03-27

 * fixed up documentation

1.00 2001-03-24

 * first version


lib/Math/BigInt/Constant.pm
#!perl

package Math::BigInt::Constant;

use strict;
use warnings;

our $VERSION = '1.12';

use Math::BigInt '1.999802';
our @ISA = qw( Math::BigInt );

use overload;                   # inherit from Math::BigInt

##############################################################################
# We Are a True Math::BigInt, But Thou Shallst Not Modify Us

sub modify {
    my ($self, $method) = @_;
    die("Can not modify ", ref($self), " $self via $method()\n");
}

##############################################################################
# But cloning us creates a modifyable Math::BigInt, so that overload works

sub copy {
    my $x = shift;

    return Math::BigInt->new($x) unless ref($x);
    Math::BigInt->copy($x);
}

1;

__END__

=pod

=head1 NAME

Math::BigInt::Constant - arbitrary sized constant integers

=head1 SYNOPSIS

  use Math::BigInt::Constant;

  my $class = 'Math::BigInt::Constant';

  # Constant creation
  $x     = $class->new($str);   # defaults to 0
  $nan   = $class->bnan();      # create a NotANumber
  $zero  = $class->bzero();     # create a "0"
  $one   = $class->bone();      # create a "1"
  $m_one = $class->bone('-');   # create a "-1"

  # Testing
  $x->is_zero();                # return wether arg is zero or not
  $x->is_nan();                 # return wether arg is NaN or not
  $x->is_one();                 # return true if arg is +1
  $x->is_one('-');              # return true if arg is -1
  $x->is_odd();                 # return true if odd, false for even
  $x->is_even();                # return true if even, false for odd
  $x->is_inf($sign);            # return true if argument is +inf or -inf, give
                                # argument ('+' or '-') to match only same sign
  $x->is_pos();                 # return true if arg > 0
  $x->is_neg();                 # return true if arg < 0

  $x->bcmp($y);                 # compare numbers (undef,<0,=0,>0)
  $x->bacmp($y);                # compare absolutely (undef,<0,=0,>0)
  $x->sign();                   # return the sign, one of +,-,+inf,-inf or NaN

  # The following would modify and thus are illegal, e.g. result in a die():

  # set
  $x->bzero();                  # set $x to 0
  $x->bnan();                   # set $x to NaN

  $x->bneg();                   # negation
  $x->babs();                   # absolute value
  $x->bnorm();                  # normalize (no-op)
  $x->bnot();                   # two's complement (bit wise not)
  $x->binc();                   # increment x by 1
  $x->bdec();                   # decrement x by 1

  $x->badd($y);                 # addition (add $y to $x)
  $x->bsub($y);                 # subtraction (subtract $y from $x)
  $x->bmul($y);                 # multiplication (multiply $x by $y)
  $x->bdiv($y);                 # divide, set $x to quotient
                                # return (quo,rem) or quo if scalar

  $x->bmod($y);                 # modulus (x % y)
  $x->bpow($y);                 # power of arguments (x ** y)
  $x->blsft($y);                # left shift
  $x->brsft($y);                # right shift

  $x->band($y);                 # bit-wise and
  $x->bior($y);                 # bit-wise inclusive or
  $x->bxor($y);                 # bit-wise exclusive or
  $x->bnot();                   # bit-wise not (two's complement)

  $x->bnok($k);                 # n over k
  $x->bfac();                   # factorial $x!
  $x->bexp();                   # Euler's number e ** $x

  $x->bsqrt();                  # calculate square-root
  $x->broot($y);                # calculate $y's root
  $x->blog($base);              # calculate integer logarithm

  $x->round($A,$P,$round_mode); # round to accuracy or precision using mode $r
  $x->bround($N);               # accuracy: preserve $N digits
  $x->bfround($N);              # round to $Nth digit, no-op for BigInts

  # The following do not modify their arguments in BigInt, so they are allowed:
  $x->bfloor();                 # return integer less or equal than $x
  $x->bceil();                  # return integer greater or equal than $x

  bgcd(@values);                # greatest common divisor
  blcm(@values);                # lowest common multiplicator

  $x->bstr();                   # return normalized string
  $x->bsstr();                  # return string in scientific notation
  $x->length();                 # return number of digits in number
  $x->digit($n);                # extract N'th digit from number

  $x->as_int();                 # return a copy of the object as BigInt
  $x->as_hex();                 # return number as hex string
  $x->as_bin();                 # return number as binary string
  $x->as_oct();                 # return number as octal string

=head1 DESCRIPTION

With this module you can define constant BigInts on a per-object basis. The
usual C<use Math::BigInt ':constant'> will catch B<all> integer constants
in the script at compile time, but will not let you create constant values
on the fly, nor work for strings and/or floating point constants like C<1e5>.

C<Math::BigInt::Constant> is a true subclass of L<Math::BigInt> and can do all
the same things - except modifying any of the objects.

=head1 EXAMPLES

Opposed to compile-time checking via C<use constant>:

    use Math::BigInt;
    use constant X => Math::BigInt->new("12345678");

    print X," ",X+2,"\n";       # okay
    print "X\n";                # oups
    X += 2;                     # not okay, will die

these provide runtime checks and can be interpolated into strings:

    use Math::BigInt::Constant;
    $x = Math::BigInt::Constant->new("3141592");

    print "$x\n";               # okay
    print $x+2,"\n";            # dito
    $x += 2;                    # not okay, will die

=head1 METHODS

A C<Math::BigInt::Constant> object has all the same methods as a
C<Math::BigInt> object.

=head1 BUGS

None discovered yet.

=head1 SEE ALSO

L<Math::BigInt>, L<Math::BigFloat::Constant>.

=head1 AUTHORS

=over 4

=item *

Tels L<http://bloodgate.com/> in early 2001-2007.

=item *

Peter John Acklam E<lt>pjacklam@online.noE<gt>, 2016-.

=back

=cut


Makefile.PL
#!perl

use strict;
use warnings;
use ExtUtils::MakeMaker 6.58;

WriteMakefile
  (
   'NAME'               => 'Math::BigInt::Constant',
   'VERSION_FROM'       => 'lib/Math/BigInt/Constant.pm',
   'AUTHOR'             => 'Peter John Acklam <pjacklam@gmail.com>',
   'MIN_PERL_VERSION'   => 5.006001,

   # A hash of modules that are required to run Makefile.PL itself, but not to
   # run the distribution.

   'CONFIGURE_REQUIRES' => {
                            'ExtUtils::MakeMaker' => 6.58,
                           },

   # A hash of modules that are needed to build the module but not run it.

   'BUILD_REQUIRES'     => {},

   # A hash of modules that are needed to run the module.

   'PREREQ_PM'          => {
                            'Math::BigInt'        => 1.999804,
                           },

   # A hash of modules that are needed to test the module but not run or build
   # it.

   'TEST_REQUIRES'      => {
                            'Test::More'          => 0.88,
                           },

   'LICENSE'            => 'perl_5',
   'LIBS'               => [''],        # e.g., '-lm'
   'DEFINE'             => '',          # e.g., '-DHAVE_SOMETHING'
   'INC'                => '',          # e.g., '-I/usr/include/other'
   'SIGN'               => 1,

   'META_MERGE'         => {
       'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/pjacklam/p5-Math-BigInt-Constant.git',
                web  => 'https://github.com/pjacklam/p5-Math-BigInt-Constant',
            },
        },
    },
  );
